plugins/mediatek-scaler/fu-mediatek-scaler-device.c

```diff
@@ -0,0 +1,1063 @@
+				     0x0,
+				     ddc_msgbox_write->data,
+				     ddc_msgbox_write->len,
+				     error);
+}
+
+static gboolean
+mediatek_scaler_device_ddc_read(FuMediatekScalerDevice *self,
+				GByteArray *buf_in,
+				GByteArray *buf_out,
+				GError **error)
+{
+	guint8 buf[0x40] = {0x00}; /* default 64 bytes */
+	g_autoptr(GBytes) rawbytes = NULL;
+	gsize report_data_sz = 0;
+	guint8 checksum = 0;
+
+	/* write for read */
+	if (!mediatek_scaler_device_ddc_write(self, buf_in, error))
+		return FALSE;
+
+	/* DDCCI spec requires host to wait at least 50 - 200ms before next message */
+	fu_device_sleep(FU_DEVICE(self), FU_MEDIATEK_SCALER_DDC_MSG_DELAY_MS);
+
+	/* read into tmp buffer */
+	if (!fu_udev_device_pread(FU_UDEV_DEVICE(self->i2c_dev), 0x0, buf, sizeof(buf), error))
+		return FALSE;
+
+	/* read buffer = addr(src) + length + data + checksum */
+	rawbytes = g_bytes_new(buf, sizeof(buf));
+	fu_dump_bytes(G_LOG_DOMAIN, "DDC/CI read buffer", rawbytes);
+
+	/* verify read buffer: [0] == source address */
+	if (buf[0] != DDC_DISPLAY_I2C_ADDR) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid read buffer: addr(src) expected 0x%02x, got 0x%02x.",
+			    (guint)DDC_DISPLAY_I2C_ADDR,
+			    buf[0]);
+		return FALSE;
+	}
+
+	/* verify read buffer: [1] as the length of data */
+	if (buf[1] <= DDC_DATA_LEN_DFT) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid read buffer: size 0x%02x must greater than 0x%02x.",
+			    buf[1],
+			    (guint)DDC_DATA_LEN_DFT);
+		return FALSE;
+	}
+
+	/* verify read buffer: overflow guard from the length of data */
+	report_data_sz = buf[1] - DDC_DATA_LEN_DFT;
+	if (report_data_sz + 3 > sizeof(buf)) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid read buffer: size 0x%02x exceeded 0x%02lx.",
+			    (guint)report_data_sz,
+			    sizeof(buf));
+		return FALSE;
+	}
+
+	/* verify read buffer: match the checksum */
+	checksum ^= DDC_HOST_I2C_ADDR_CHKSUM;
+	for (gsize i = 0; i < report_data_sz + 2; i++)
+		checksum ^= buf[i];
+
+	if (buf[report_data_sz + 2] != checksum) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid read buffer, checksum expected 0x%02x, got 0x%02x.",
+			    checksum,
+			    buf[report_data_sz + 2]);
+		return FALSE;
+	}
+
+	/* checksum is needn't */
+	g_byte_array_append(buf_out, buf, report_data_sz + 2);
+
+	/* print the raw data */
+	rawbytes = g_bytes_new(buf_out->data, (gsize)buf_out->len);
+	fu_dump_bytes(G_LOG_DOMAIN, "DDC/CI read report", rawbytes);
+	return TRUE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_set_ddc_priority(FuDevice *device, guint8 flag, GError **error)
+{
+	FuMediatekScalerDevice *self = FU_MEDIATEK_SCALER_DEVICE(device);
+	g_autoptr(GByteArray) cod = g_byte_array_new();
+	g_autoptr(GError) error_local = NULL;
+	const guint8 buf[2] = {DDC_OPCODE_MTK_REQ, DDC_VCP_CODE_PRIORITY};
```

Comment:
i think these would be much clearer as structs in the .rs file -- can i push to your branch with an example?

Reply:
Sure, please. 👍

Reply:
you have a try first -- and if you get stuck yell. See https://fwupd.github.io/libfwupdplugin/tutorial.html#the-rustgen-helper for docs

Reply:
happy with the enum objects, haven't got that far for code blocks (seems more complicated). Would you like to make changes or we can let it go?

Reply:
I've made some changes -- can you verify I've not broken anything please :)

Reply:
Amazing!! Okay, will test it out next Mon.

Reply:
Those changes taste sweet. :) Thanks brighten up the way to rust uses.

Reply:
How about let's merge the two commits into one?